// @ks89/angular-modal-gallery/esm2015/ks89-angular-modal-gallery.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { GalleryModule, Action, Image, ImageEvent, ImageModalEvent, DescriptionStrategy, ButtonsStrategy, ButtonType, GalleryService, LoadingType, KS_DEFAULT_ACCESSIBILITY_CONFIG, KS_DEFAULT_BTN_FULL_SCREEN, KS_DEFAULT_BTN_CLOSE, KS_DEFAULT_BTN_DELETE, KS_DEFAULT_BTN_DOWNLOAD, KS_DEFAULT_BTN_EXTURL, KS_DEFAULT_SIZE, LineLayout, GridLayout, AdvancedLayout, PlainGalleryStrategy } from './public_api';
export { AccessibleComponent as ɵh } from './lib/components/accessible.component';
export { BackgroundComponent as ɵe } from './lib/components/background/background.component';
export { CarouselPreviewsComponent as ɵm } from './lib/components/carousel/carousel-previews/carousel-previews.component';
export { CarouselComponent as ɵd } from './lib/components/carousel/carousel.component';
export { COMPONENTS as ɵb } from './lib/components/components';
export { CurrentImageComponent as ɵg } from './lib/components/current-image/current-image.component';
export { LoadingSpinnerComponent as ɵq } from './lib/components/current-image/loading-spinner/loading-spinner.component';
export { DotsComponent as ɵo } from './lib/components/dots/dots.component';
export { ModalGalleryComponent as ɵc } from './lib/components/modal-gallery/modal-gallery.component';
export { PlainGalleryComponent as ɵf } from './lib/components/plain-gallery/plain-gallery.component';
export { PreviewsComponent as ɵp } from './lib/components/previews/previews.component';
export { UpperButtonsComponent as ɵn } from './lib/components/upper-buttons/upper-buttons.component';
export { ATagBgImageDirective as ɵx } from './lib/directives/a-tag-bg-image.directive';
export { ClickOutsideDirective as ɵs } from './lib/directives/click-outside.directive';
export { DescriptionDirective as ɵy } from './lib/directives/description.directive';
export { DirectionDirective as ɵw } from './lib/directives/direction.directive';
export { DIRECTIVES as ɵr } from './lib/directives/directives';
export { KeyboardNavigationDirective as ɵu } from './lib/directives/keyboard-navigation.directive';
export { MarginDirective as ɵz } from './lib/directives/margin.directive';
export { MaxSizeDirective as ɵba } from './lib/directives/max-size.directive';
export { SizeDirective as ɵt } from './lib/directives/size.directive';
export { WrapDirective as ɵv } from './lib/directives/wrap.directive';
export { setupKeyboardService as ɵa } from './lib/modal-gallery.module';
export { IdValidatorService as ɵl } from './lib/services/id-validator.service';
export { KEYBOARD_CONFIGURATION as ɵi, KeyboardService as ɵj } from './lib/services/keyboard.service';
