// angular-tilt/esm2015/lib/angular-tilt.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from "@angular/core";
import { VanillaTilt } from "./angular-tilt";
export class AngularTiltDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        this.el = el;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.tilt = new VanillaTilt(this.el.nativeElement, this.tiltSettings);
    }
}
AngularTiltDirective.decorators = [
    { type: Directive, args: [{
                selector: "[aTilt]"
            },] }
];
/** @nocollapse */
AngularTiltDirective.ctorParameters = () => [
    { type: ElementRef }
];
AngularTiltDirective.propDecorators = {
    tiltSettings: [{ type: Input, args: ["tiltSettings",] }]
};
if (false) {
    /** @type {?} */
    AngularTiltDirective.prototype.tilt;
    /** @type {?} */
    AngularTiltDirective.prototype.tiltSettings;
    /**
     * @type {?}
     * @private
     */
    AngularTiltDirective.prototype.el;
}
