// angular-tilt/esm5/lib/angular-tilt.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input } from "@angular/core";
import { VanillaTilt } from "./angular-tilt";
var AngularTiltDirective = /** @class */ (function () {
    function AngularTiltDirective(el) {
        this.el = el;
    }
    /**
     * @return {?}
     */
    AngularTiltDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.tilt = new VanillaTilt(this.el.nativeElement, this.tiltSettings);
    };
    AngularTiltDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[aTilt]"
                },] }
    ];
    /** @nocollapse */
    AngularTiltDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    AngularTiltDirective.propDecorators = {
        tiltSettings: [{ type: Input, args: ["tiltSettings",] }]
    };
    return AngularTiltDirective;
}());
export { AngularTiltDirective };
if (false) {
    /** @type {?} */
    AngularTiltDirective.prototype.tilt;
    /** @type {?} */
    AngularTiltDirective.prototype.tiltSettings;
    /**
     * @type {?}
     * @private
     */
    AngularTiltDirective.prototype.el;
}
