// ng2-animate-on-scroll/dist/src/scroll.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var rxjs_1 = require("rxjs");
var fromEvent_1 = require("rxjs/observable/fromEvent");
var ScrollService = (function () {
    function ScrollService() {
        var _this = this;
        this.scrollSub = new rxjs_1.Subscription();
        this.resizeSub = new rxjs_1.Subscription();
        // set initial value
        this.manageScrollPos();
        // create observable that we can subscribe to from component or directive
        this.scrollObs = fromEvent_1.fromEvent(window, 'scroll');
        // initiate subscription to update values
        this.scrollSub = this.scrollObs
            .subscribe(function () { return _this.manageScrollPos(); });
        // create observable for changes in screen size
        this.resizeObs = fromEvent_1.fromEvent(window, 'resize');
        // initiate subscription to update values
        this.resizeSub = this.resizeObs
            .subscribe(function () { return _this.manageScrollPos(); });
    }
    ScrollService.prototype.manageScrollPos = function () {
        // update service property
        this.pos = window.pageYOffset;
    };
    ScrollService.prototype.ngOnDestroy = function () {
        this.scrollSub.unsubscribe();
        this.resizeSub.unsubscribe();
    };
    ScrollService.decorators = [
        { type: core_1.Injectable },
    ];
    /** @nocollapse */
    ScrollService.ctorParameters = function () { return []; };
    return ScrollService;
}());
exports.ScrollService = ScrollService;
