// @ks89/angular-modal-gallery/esm5/lib/model/buttons-config.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
/**
 * Interface `ButtonsConfig` to add buttons, show/hide their, and to add the strategy.
 * @record
 */
export function ButtonsConfig() { }
if (false) {
    /** @type {?} */
    ButtonsConfig.prototype.visible;
    /** @type {?} */
    ButtonsConfig.prototype.strategy;
    /** @type {?|undefined} */
    ButtonsConfig.prototype.buttons;
}
/**
 * Interface `ButtonConfig` to configure a single button.
 * @record
 */
export function ButtonConfig() { }
if (false) {
    /** @type {?|undefined} */
    ButtonConfig.prototype.className;
    /** @type {?|undefined} */
    ButtonConfig.prototype.size;
    /** @type {?|undefined} */
    ButtonConfig.prototype.fontSize;
    /** @type {?} */
    ButtonConfig.prototype.type;
    /** @type {?|undefined} */
    ButtonConfig.prototype.title;
    /** @type {?|undefined} */
    ButtonConfig.prototype.ariaLabel;
    /** @type {?|undefined} */
    ButtonConfig.prototype.extUrlInNewTab;
}
/**
 * Interface `ButtonEvent` to represent the event payload when a button is clicked.
 * @record
 */
export function ButtonEvent() { }
if (false) {
    /** @type {?} */
    ButtonEvent.prototype.button;
    /** @type {?} */
    ButtonEvent.prototype.image;
    /** @type {?} */
    ButtonEvent.prototype.action;
}
/** @enum {number} */
var ButtonsStrategy = {
    // don't use 0 here
    // the first index is 1 and all of the following members are auto-incremented from that point on
    DEFAULT: 1,
    SIMPLE: 2,
    ADVANCED: 3,
    FULL: 4,
    CUSTOM: 5,
};
export { ButtonsStrategy };
ButtonsStrategy[ButtonsStrategy.DEFAULT] = 'DEFAULT';
ButtonsStrategy[ButtonsStrategy.SIMPLE] = 'SIMPLE';
ButtonsStrategy[ButtonsStrategy.ADVANCED] = 'ADVANCED';
ButtonsStrategy[ButtonsStrategy.FULL] = 'FULL';
ButtonsStrategy[ButtonsStrategy.CUSTOM] = 'CUSTOM';
/** @enum {number} */
var ButtonType = {
    // don't use 0 here
    // the first index is 1 and all of the following members are auto-incremented from that point on
    // REFRESH = 1,
    DELETE: 1,
    EXTURL: 2,
    DOWNLOAD: 3,
    CLOSE: 4,
    CUSTOM: 5,
    FULLSCREEN: 6,
    // ROTATE
};
export { ButtonType };
ButtonType[ButtonType.DELETE] = 'DELETE';
ButtonType[ButtonType.EXTURL] = 'EXTURL';
ButtonType[ButtonType.DOWNLOAD] = 'DOWNLOAD';
ButtonType[ButtonType.CLOSE] = 'CLOSE';
ButtonType[ButtonType.CUSTOM] = 'CUSTOM';
ButtonType[ButtonType.FULLSCREEN] = 'FULLSCREEN';
/**
 * Array of admitted types of buttons.
 * @type {?}
 */
export var WHITELIST_BUTTON_TYPES = [
    // ButtonType.REFRESH,
    ButtonType.FULLSCREEN,
    ButtonType.DELETE,
    ButtonType.EXTURL,
    ButtonType.DOWNLOAD,
    ButtonType.CLOSE,
    ButtonType.CUSTOM
    // ButtonType.ROTATE
];
