// @ks89/angular-modal-gallery/esm5/lib/model/image.class.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
/**
 * Class `Image` that represents an image with both `modal` and `plain` configurations.
 * Both image `id` and `modal` are mandatory, instead `plain` is optional.
 */
var /**
 * Class `Image` that represents an image with both `modal` and `plain` configurations.
 * Both image `id` and `modal` are mandatory, instead `plain` is optional.
 */
Image = /** @class */ (function () {
    function Image(id, modal, plain) {
        this.id = id;
        this.modal = modal;
        this.plain = plain;
    }
    return Image;
}());
/**
 * Class `Image` that represents an image with both `modal` and `plain` configurations.
 * Both image `id` and `modal` are mandatory, instead `plain` is optional.
 */
export { Image };
if (false) {
    /** @type {?} */
    Image.prototype.id;
    /** @type {?} */
    Image.prototype.modal;
    /** @type {?} */
    Image.prototype.plain;
}
/**
 * Interface `ImageData` to configure an image, but it isn't used directly.
 * Please, refers to `PlainImage` or `ModalImage`.
 * @record
 */
export function ImageData() { }
if (false) {
    /** @type {?} */
    ImageData.prototype.img;
    /** @type {?|undefined} */
    ImageData.prototype.description;
    /** @type {?|undefined} */
    ImageData.prototype.title;
    /** @type {?|undefined} */
    ImageData.prototype.alt;
    /** @type {?|undefined} */
    ImageData.prototype.ariaLabel;
    /** @type {?|undefined} */
    ImageData.prototype.angle;
}
/**
 * Interface `ModalImage` to configure the modal image.
 * @record
 */
export function ModalImage() { }
if (false) {
    /** @type {?|undefined} */
    ModalImage.prototype.extUrl;
    /** @type {?|undefined} */
    ModalImage.prototype.downloadFileName;
}
/**
 * Interface `PlainImage` to configure the plain image.
 * @record
 */
export function PlainImage() { }
if (false) {
    /** @type {?|undefined} */
    PlainImage.prototype.size;
}
/**
 * Class `ImageEvent` that represents the event payload with the result and the triggered action.
 */
var /**
 * Class `ImageEvent` that represents the event payload with the result and the triggered action.
 */
ImageEvent = /** @class */ (function () {
    function ImageEvent(action, result) {
        this.action = action;
        this.result = result;
    }
    return ImageEvent;
}());
/**
 * Class `ImageEvent` that represents the event payload with the result and the triggered action.
 */
export { ImageEvent };
if (false) {
    /** @type {?} */
    ImageEvent.prototype.action;
    /** @type {?} */
    ImageEvent.prototype.result;
}
/**
 * Class `ImageModalEvent` that represents the event payload with the result and the triggered action.
 */
var /**
 * Class `ImageModalEvent` that represents the event payload with the result and the triggered action.
 */
ImageModalEvent = /** @class */ (function (_super) {
    tslib_1.__extends(ImageModalEvent, _super);
    function ImageModalEvent(action, result) {
        return _super.call(this, action, result) || this;
    }
    return ImageModalEvent;
}(ImageEvent));
/**
 * Class `ImageModalEvent` that represents the event payload with the result and the triggered action.
 */
export { ImageModalEvent };
