// @ks89/angular-modal-gallery/esm5/lib/model/plain-gallery-config.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
/**
 * Interface `PlainGalleryConfig` to configure plain-gallery features.
 * @record
 */
export function PlainGalleryConfig() { }
if (false) {
    /** @type {?} */
    PlainGalleryConfig.prototype.strategy;
    /** @type {?} */
    PlainGalleryConfig.prototype.layout;
    /** @type {?|undefined} */
    PlainGalleryConfig.prototype.advanced;
}
/**
 * Interface `PlainGalleryLayout` to configure the layout. This interface isn't used directly, instead
 * refers to either `LineLayout`, `GridLayout` or `AdvancedLayout`.
 * @record
 */
export function PlainGalleryLayout() { }
/**
 * Class `LineLayout` to configure a linear plain gallery.
 */
var /**
 * Class `LineLayout` to configure a linear plain gallery.
 */
LineLayout = /** @class */ (function () {
    function LineLayout(size, breakConfig, justify) {
        this.size = size;
        this.breakConfig = breakConfig;
        this.justify = justify;
    }
    return LineLayout;
}());
/**
 * Class `LineLayout` to configure a linear plain gallery.
 */
export { LineLayout };
if (false) {
    /** @type {?} */
    LineLayout.prototype.breakConfig;
    /** @type {?} */
    LineLayout.prototype.justify;
    /** @type {?} */
    LineLayout.prototype.size;
}
/**
 * Class `GridLayout` to configure a grid plain gallery.
 */
var /**
 * Class `GridLayout` to configure a grid plain gallery.
 */
GridLayout = /** @class */ (function () {
    function GridLayout(size, breakConfig) {
        this.size = size;
        this.breakConfig = breakConfig;
    }
    return GridLayout;
}());
/**
 * Class `GridLayout` to configure a grid plain gallery.
 */
export { GridLayout };
if (false) {
    /** @type {?} */
    GridLayout.prototype.breakConfig;
    /** @type {?} */
    GridLayout.prototype.size;
}
/**
 * Class `AdvancedLayout` to configure a fully custom plain gallery.
 */
var /**
 * Class `AdvancedLayout` to configure a fully custom plain gallery.
 */
AdvancedLayout = /** @class */ (function () {
    function AdvancedLayout(modalOpenerByIndex, hideDefaultPlainGallery) {
        this.modalOpenerByIndex = modalOpenerByIndex;
        this.hideDefaultPlainGallery = hideDefaultPlainGallery;
    }
    return AdvancedLayout;
}());
/**
 * Class `AdvancedLayout` to configure a fully custom plain gallery.
 */
export { AdvancedLayout };
if (false) {
    /** @type {?} */
    AdvancedLayout.prototype.modalOpenerByIndex;
    /** @type {?} */
    AdvancedLayout.prototype.hideDefaultPlainGallery;
}
/** @enum {number} */
var PlainGalleryStrategy = {
    // don't use 0 here
    // the first index is 1 and all of the following members are auto-incremented from that point on
    ROW: 1,
    COLUMN: 2,
    GRID: 3,
    CUSTOM: 4 // full custom strategy
    ,
};
export { PlainGalleryStrategy };
PlainGalleryStrategy[PlainGalleryStrategy.ROW] = 'ROW';
PlainGalleryStrategy[PlainGalleryStrategy.COLUMN] = 'COLUMN';
PlainGalleryStrategy[PlainGalleryStrategy.GRID] = 'GRID';
PlainGalleryStrategy[PlainGalleryStrategy.CUSTOM] = 'CUSTOM';
/**
 * Interface `BreakConfig` to limit the number of items of the plain gallery or to force it to fill other lines.
 * @record
 */
export function BreakConfig() { }
if (false) {
    /** @type {?} */
    BreakConfig.prototype.length;
    /** @type {?} */
    BreakConfig.prototype.wrap;
}
/**
 * Interface `AdvancedConfig` to use `<a>` tags instead of `<img>`.
 * It also contains a string property to customize the css background property.
 * For more info check here https://www.w3schools.com/cssref/css3_pr_background.asp
 * @record
 */
export function AdvancedConfig() { }
if (false) {
    /** @type {?} */
    AdvancedConfig.prototype.aTags;
    /** @type {?} */
    AdvancedConfig.prototype.additionalBackground;
}
