// @ks89/angular-modal-gallery/esm5/lib/services/gallery.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { EventEmitter, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function InternalGalleryPayload() { }
if (false) {
    /** @type {?} */
    InternalGalleryPayload.prototype.galleryId;
    /** @type {?|undefined} */
    InternalGalleryPayload.prototype.index;
    /** @type {?|undefined} */
    InternalGalleryPayload.prototype.image;
    /** @type {?|undefined} */
    InternalGalleryPayload.prototype.result;
}
/**
 * Service to call methods on a gallery by its galleryId.
 */
var GalleryService = /** @class */ (function () {
    function GalleryService() {
        this.navigate = new EventEmitter();
        this.close = new EventEmitter();
        this.update = new EventEmitter();
        this.autoPlay = new EventEmitter();
    }
    /**
     * Method to open the modal gallery with the galleryId passed as parameter.
     * It will automatically shows the image at position index.
     * @param galleryId number or undefined that represents the unique id of the gallery.
     * @param index number of the image that you want to open.
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     */
    /**
     * Method to open the modal gallery with the galleryId passed as parameter.
     * It will automatically shows the image at position index.
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @param {?} index number of the image that you want to open.
     * @return {?}
     */
    GalleryService.prototype.openGallery = /**
     * Method to open the modal gallery with the galleryId passed as parameter.
     * It will automatically shows the image at position index.
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @param {?} index number of the image that you want to open.
     * @return {?}
     */
    function (galleryId, index) {
        if (galleryId === undefined || galleryId < 0 || index < 0) {
            throw new Error('Cannot open gallery via GalleryService with either index<0 or galleryId<0 or galleryId===undefined');
        }
        this.navigate.emit({
            galleryId: galleryId,
            index: index
        });
    };
    /**
     * Method to navigate to a specific index of the modal gallery with the galleryId passed as parameter.
     * At the moment, it's like openGallery, but in upcoming releases it will change the behaviour.
     * @param galleryId number or undefined that represents the unique id of the gallery.
     * @param index number of the image that you want to open.
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     */
    /**
     * Method to navigate to a specific index of the modal gallery with the galleryId passed as parameter.
     * At the moment, it's like openGallery, but in upcoming releases it will change the behaviour.
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @param {?} index number of the image that you want to open.
     * @return {?}
     */
    GalleryService.prototype.navigateGallery = /**
     * Method to navigate to a specific index of the modal gallery with the galleryId passed as parameter.
     * At the moment, it's like openGallery, but in upcoming releases it will change the behaviour.
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @param {?} index number of the image that you want to open.
     * @return {?}
     */
    function (galleryId, index) {
        if (galleryId === undefined || galleryId < 0 || index < 0) {
            throw new Error('Cannot navigate via GalleryService with either index<0 or galleryId<0 or galleryId===undefined');
        }
        this.navigate.emit({
            galleryId: galleryId,
            index: index
        });
    };
    /**
     * Method to close the modal gallery with the galleryId passed as parameter.
     * @param galleryId number or undefined that represents the unique id of the gallery.
     * @throws a error with a message if galleryId is either undefined or < 0
     */
    /**
     * Method to close the modal gallery with the galleryId passed as parameter.
     * @throws a error with a message if galleryId is either undefined or < 0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @return {?}
     */
    GalleryService.prototype.closeGallery = /**
     * Method to close the modal gallery with the galleryId passed as parameter.
     * @throws a error with a message if galleryId is either undefined or < 0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @return {?}
     */
    function (galleryId) {
        if (galleryId === undefined || galleryId < 0) {
            throw new Error('Cannot close gallery via GalleryService with galleryId<0 or galleryId===undefined');
        }
        this.close.emit(galleryId);
    };
    /**
     * Service to update an image with a new object
     * @param galleryId number or undefined that represents the unique id of the gallery.
     * @param index number of the image that you want to update.
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     *
     * @since 6.3.0
     */
    /**
     * Service to update an image with a new object
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     *
     * \@since 6.3.0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @param {?} index number of the image that you want to update.
     * @param {?} image
     * @return {?}
     */
    GalleryService.prototype.updateGallery = /**
     * Service to update an image with a new object
     * @throws a error with a message if galleryId is either undefined, < 0 or index is < 0
     *
     * \@since 6.3.0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @param {?} index number of the image that you want to update.
     * @param {?} image
     * @return {?}
     */
    function (galleryId, index, image) {
        if (galleryId === undefined || galleryId < 0 || index < 0) {
            throw new Error('Cannot update gallery via GalleryService with either index<0 or galleryId<0 or galleryId===undefined');
        }
        if (!image) {
            throw new Error('Cannot update gallery via GalleryService, because image is not valid');
        }
        this.update.emit({
            galleryId: galleryId,
            index: index,
            image: image
        });
    };
    /**
     * Service to play modal-gallery
     * @param galleryId number or undefined that represents the unique id of the gallery.
     * @throws a error with a message if galleryId is either undefined or < 0
     *
     * @since 7.2.0
     */
    /**
     * Service to play modal-gallery
     * @throws a error with a message if galleryId is either undefined or < 0
     *
     * \@since 7.2.0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @return {?}
     */
    GalleryService.prototype.play = /**
     * Service to play modal-gallery
     * @throws a error with a message if galleryId is either undefined or < 0
     *
     * \@since 7.2.0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @return {?}
     */
    function (galleryId) {
        if (galleryId === undefined || galleryId < 0) {
            throw new Error('Cannot play gallery via GalleryService with galleryId<0 or galleryId===undefined');
        }
        this.autoPlay.emit({
            galleryId: galleryId,
            result: true
        });
    };
    /**
     * Service to stop modal-gallery
     * @param galleryId number or undefined that represents the unique id of the gallery.
     * @throws a error with a message if galleryId is either undefined or < 0
     *
     * @since 7.2.0
     */
    /**
     * Service to stop modal-gallery
     * @throws a error with a message if galleryId is either undefined or < 0
     *
     * \@since 7.2.0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @return {?}
     */
    GalleryService.prototype.stop = /**
     * Service to stop modal-gallery
     * @throws a error with a message if galleryId is either undefined or < 0
     *
     * \@since 7.2.0
     * @param {?} galleryId number or undefined that represents the unique id of the gallery.
     * @return {?}
     */
    function (galleryId) {
        if (galleryId === undefined || galleryId < 0) {
            throw new Error('Cannot stop gallery via GalleryService with galleryId<0 or galleryId===undefined');
        }
        this.autoPlay.emit({
            galleryId: galleryId,
            result: false
        });
    };
    GalleryService.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */ GalleryService.ngInjectableDef = i0.defineInjectable({ factory: function GalleryService_Factory() { return new GalleryService(); }, token: GalleryService, providedIn: "root" });
    return GalleryService;
}());
export { GalleryService };
if (false) {
    /** @type {?} */
    GalleryService.prototype.navigate;
    /** @type {?} */
    GalleryService.prototype.close;
    /** @type {?} */
    GalleryService.prototype.update;
    /** @type {?} */
    GalleryService.prototype.autoPlay;
}
